Kernels compilation profiling doc (#4765)
intel/intel-xpu-backend-for-triton: docs/programming-guide/chapter-3/profiling.rst

=============================
Profiling kernels compilation
=============================

> If a kernel takes too much time to compile, the first step to start profiling it 

Reviewer: ```suggestion
If a kernel takes too much time to compile, the first step to start profiling it
```

> If some kernel takes to much time to compile, the first step to start the profiling with
> is to run the program with the following environment variables:

Reviewer: ```suggestion
If some kernel takes to much time to compile, the first step to start the profiling by
running the program with the following environment variables:
```

> If some kernel takes to much time to compile, the first step to start the profiling with
> is to run the program with the following environment variables:

Reviewer 2: ```suggestion
If a kernel takes too much time to compile, the first step to start profiling it 
is to run it with the following environment variables:
```

.. code-block:: bash

    export TRITON_ALWAYS_COMPILE=1
    export MLIR_ENABLE_TIMING=1
    export LLVM_ENABLE_TIMING=1
    python -m pytest python/test/unit/intel/test_mxfp_matmul.py::test_mxfp_matmul[True-True-float4-float4-True-True-1-128-128-128-1024-512-512] --device=xpu -s

> For the variables description, see the README.md file in the root directory of the project.

Reviewer: ```suggestion
For the environment variables description, see the README.md file in the root directory of the project.
```

        0.0011 (  0.0%)    0.0011 (  0.0%)  TritonGPUGlobalScratchAllocationPass
        0.5947 (  0.1%)    0.5947 (  0.1%)  ConvertTritonIntelGPUToLLVM
        0.1232 (  0.0%)    0.1232 (  0.0%)  ConvertTritonGENToLLVM
        0.3430 (  0.1%)    0.3430 (  0.1%)  TritonIntelGPURewriteStackPtr
        15.3707 (  3.1%)   15.3707 (  3.2%)  Canonicalizer
        0.1650 (  0.0%)    0.1650 (  0.0%)  CSE
        0.0000 (  0.0%)    0.0000 (  0.0%)    (A) DominanceInfo
        0.0775 (  0.0%)    0.0775 (  0.0%)  ArithToLLVMConversionPass
        0.1934 (  0.0%)    0.1934 (  0.0%)  Canonicalizer
        0.1046 (  0.0%)    0.1046 (  0.0%)  CSE
        0.0000 (  0.0%)    0.0000 (  0.0%)    (A) DominanceInfo
        0.1470 (  0.0%)    0.1470 (  0.0%)  SymbolDCE
        0.0876 (  0.0%)    0.0876 (  0.0%)  LLVMDIScope
        483.1849 ( 96.6%)  465.9228 ( 96.4%)  Rest
        500.4471 (100.0%)  483.1849 (100.0%)  Total
            49 function calls in 465.865 seconds

What we can see here is that the most time-consuming part of the compilation is something,
which is called "Rest" and we don't know what it is. For further investigation, we can

> use the Python's `cProfiler`. To profile not the entire program, but only individual functions,

Reviewer 2: ```suggestion
use the Python's `cProfiler`. To profile individual functions,
```

we can use the following decorator:

.. code-block:: python

    def profile(func):
        import cProfile
        import pstats
        import functools

        @functools.wraps(func)
        def wrapper(*args, **kwargs):
            print(f"Profiling {func.__qualname__}")
            pr = cProfile.Profile()
            pr.enable()
            try:
                return func(*args, **kwargs)
            finally:
                pr.disable()
                ps = pstats.Stats(pr).sort_stats('cumulative')
                ps.print_stats(20)
        return wrapper

After decorating the `make_llir` function:

.. code-block:: python

    @profile
    def make_llir(...):
        ...

and running the program again, we will get the following results:

.. code-block:: text

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1  465.552  465.552  465.552  465.552 {built-in method triton._C.libtriton.intel.optimize_module}
        1    0.231    0.231    0.231    0.231 {built-in method triton._C.libtriton.llvm.to_module}
        ...

> Now we see that the most time is spent in the `optimize_module` function, which is a native function,

Reviewer 2: ```suggestion
The new profile information indicates that most of the time is spent in the `optimize_module` function, which is a native function,
```
